File: lattice_brain/graph/provenance.py
from __future__ import annotations

from typing import TYPE_CHECKING

# ruff: noqa: F403,F405
from ._kg_common import *  # noqa: F403,F401

# The cross-mixin surface (`_connect`, `_upsert_node`, …) is declared in
# `_kg_contract.KnowledgeGraphCore`. It is a typing-only base: at runtime this
# is `object`, so the MRO of `KnowledgeGraphStore` is unchanged.
if TYPE_CHECKING:
    from ._kg_contract import KnowledgeGraphCore as _Core
else:
    _Core = object



class KnowledgeGraphProvenanceMixin(_Core):
    def record_provenance(
        self,
        *,
        node_id: str,
        source_type: str,
        pipeline: str = "unified-ingestion",
        source_uri: Optional[str] = None,
        content_hash: Optional[str] = None,
        title: Optional[str] = None,
        owner: Optional[str] = None,
        workspace_id: Optional[str] = None,
        captured_at: Optional[str] = None,
        modified_at: Optional[str] = None,
        embedded: bool = False,
        linked: bool = False,
        duplicate: bool = False,
        agent_used: Optional[str] = None,
        chunk_count: int = 0,
        permissions: Optional[Dict[str, Any]] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Append a provenance record for an ingested node (audit trail)."""
        now = _now()
        prov_basis = f"{node_id}|{content_hash or ''}|{now}"
        prov_id = f"prov:{_sha256_text(prov_basis)[:24]}"
        with self._connect() as conn:
            conn.execute(
                """
                    INSERT OR REPLACE INTO ingestion_provenance(
                      id, node_id, source_type, source_uri, content_hash, title, pipeline,
                      owner, workspace_id, captured_at, modified_at, embedded, linked,
                      duplicate, agent_used, chunk_count, permissions_json, metadata_json, created_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                (
                    prov_id,
                    node_id,
                    source_type,
                    source_uri,
                    content_hash,
                    title,
                    pipeline,
                    owner,
                    workspace_id,
                    captured_at,
                    modified_at,
                    1 if embedded else 0,
                    1 if linked else 0,
                    1 if duplicate else 0,
                    agent_used,
                    int(chunk_count or 0),
                    _json(permissions or {}),
                    _json(metadata or {}),
                    now,
                ),
            )
        return {"id": prov_id, "node_id": node_id, "created_at": now}

    @staticmethod
    def _provenance_row(row: sqlite3.Row) -> Dict[str, Any]:
        return {
            "id": row["id"],
            "node_id": row["node_id"],
            "source_type": row["source_type"],
            "source_uri": row["source_uri"],
            "content_hash": row["content_hash"],
            "title": row["title"],
            "pipeline": row["pipeline"],
            "owner": row["owner"],
            "workspace_id": row["workspace_id"],
            "captured_at": row["captured_at"],
            "modified_at": row["modified_at"],
            "embedded": bool(row["embedded"]),
            "linked": bool(row["linked"]),
            "duplicate": bool(row["duplicate"]),
            "agent_used": row["agent_used"],
            "chunk_count": row["chunk_count"],
            "permissions": _safe_loads(row["permissions_json"]),
            "metadata": _safe_loads(row["metadata_json"]),
            "created_at": row["created_at"],
        }

    def get_provenance(self, node_id: str) -> Optional[Dict[str, Any]]:
        """Return the most recent provenance record for a node, or None."""
        with self._connect() as conn:
            row = conn.execute(
                "SELECT * FROM ingestion_provenance WHERE node_id = ? "
                "ORDER BY created_at DESC, rowid DESC LIMIT 1",
                (node_id,),
            ).fetchone()
            return self._provenance_row(row) if row else None

    def list_provenance(
        self, *, limit: int = 100, source_type: Optional[str] = None
    ) -> Dict[str, Any]:
        """Recent provenance records (newest first), optionally by source_type."""
        limit = max(1, min(int(limit or 100), 1000))
        with self._connect() as conn:
            if source_type:
                rows = conn.execute(
                    "SELECT * FROM ingestion_provenance WHERE source_type = ? "
                    "ORDER BY created_at DESC, rowid DESC LIMIT ?",
                    (source_type, limit),
                ).fetchall()
            else:
                rows = conn.execute(
                    "SELECT * FROM ingestion_provenance "
                    "ORDER BY created_at DESC, rowid DESC LIMIT ?",
                    (limit,),
                ).fetchall()
            return {
                "items": [self._provenance_row(r) for r in rows],
                "count": len(rows),
            }

    def provenance_coverage(self) -> Dict[str, Any]:
        """How much of the brain is explainable: nodes with vs without
        provenance, per node type — the honesty metric for 'every source goes
        through the pipeline'. Pre-v4 nodes ingested before provenance existed
        legitimately count as uncovered."""
        nt, _ = self._read_tables()
        with self._connect() as conn:
            total = conn.execute(f"SELECT COUNT(*) FROM {nt}").fetchone()[0]
            covered = conn.execute(
                f"SELECT COUNT(*) FROM {nt} WHERE id IN (SELECT DISTINCT node_id FROM ingestion_provenance)"
            ).fetchone()[0]
            uncovered_by_type = {
                row["type"]: row["c"]
                for row in conn.execute(
                    f"""
                        SELECT type, COUNT(*) AS c FROM {nt}
                        WHERE id NOT IN (SELECT DISTINCT node_id FROM ingestion_provenance)
                        GROUP BY type ORDER BY c DESC LIMIT 20
                        """
                ).fetchall()
            }
            by_source = {
                row["source_type"]: row["c"]
                for row in conn.execute(
                    "SELECT source_type, COUNT(*) AS c FROM ingestion_provenance GROUP BY source_type"
                ).fetchall()
            }
        return {
            "total_nodes": total,
            "nodes_with_provenance": covered,
            "coverage_ratio": round(covered / total, 4) if total else None,
            "uncovered_by_type": uncovered_by_type,
            "provenance_by_source_type": by_source,
        }

    def provenance_stats(self) -> Dict[str, Any]:
        """Aggregate provenance counts for the Knowledge Graph status surface."""
        with self._connect() as conn:
            total = conn.execute(
                "SELECT COUNT(*) AS c FROM ingestion_provenance"
            ).fetchone()["c"]
            by_source = {
                r["source_type"]: r["c"]
                for r in conn.execute(
                    "SELECT source_type, COUNT(*) AS c FROM ingestion_provenance GROUP BY source_type"
                ).fetchall()
            }
            embedded = conn.execute(
                "SELECT COUNT(*) AS c FROM ingestion_provenance WHERE embedded = 1"
            ).fetchone()["c"]
            duplicates = conn.execute(
                "SELECT COUNT(*) AS c FROM ingestion_provenance WHERE duplicate = 1"
            ).fetchone()["c"]
            last = conn.execute(
                "SELECT created_at FROM ingestion_provenance ORDER BY created_at DESC LIMIT 1"
            ).fetchone()
        return {
            "total": total,
            "by_source_type": by_source,
            "embedded": embedded,
            "duplicates": duplicates,
            "last_ingested_at": last["created_at"] if last else None,
        }

    def schema_versions(self) -> Dict[str, Any]:
        """Versions an exporter stamps and an importer validates against."""
        try:
            from .schema import EMBED_DIM as _EMBED_DIM
            from .schema import KG_SCHEMA_V2_VERSION as _V2
        except Exception:  # pragma: no cover - kg_schema always importable in practice
            _EMBED_DIM, _V2 = 1024, 2
        return {
            "graph_schema_version": GRAPH_SCHEMA_VERSION,
            "db_format_version": _KG_DB_FORMAT_VERSION,
            "kg_v2_schema_version": _V2,
            "projection_version": _PROJECTION_VERSION,
            "embed_dim": _EMBED_DIM,
        }

    def export_graph_data(
        self,
        *,
        workspace_id: Optional[str] = None,
        include_legacy_global: bool = False,
    ) -> Dict[str, Any]:
        """Raw, lossless logical export of the graph (nodes/edges/chunks/sources/
        provenance). Vector embeddings are intentionally omitted — they are
        re-derived on import — so the artifact stays portable and small. Use
        :meth:`backup_database` for a faithful binary copy incl. embeddings.

        ``workspace_id`` REALLY filters (v4): the artifact contains only nodes
        scoped to that workspace, with edges/chunks/provenance restricted to the
        surviving nodes. Legacy-global rows require the explicit
        ``include_legacy_global=True`` migration/compatibility opt-in. Pre-v4
        this parameter was stamped into the header while the data exported
        everything — a header that lied.
        """
        with self._connect() as conn:

            def rows(table: str):
                return [
                    dict(r) for r in conn.execute(f"SELECT * FROM {table}").fetchall()
                ]

            if workspace_id:
                scope_sql = "workspace_id = ?"
                if include_legacy_global:
                    scope_sql += " OR workspace_id IS NULL"
                keep_ids = {
                    row["id"]
                    for row in conn.execute(
                        f"SELECT id FROM nodes_v2 WHERE {scope_sql}",
                        (workspace_id,),
                    ).fetchall()
                }
                nodes = [n for n in rows("nodes") if n["id"] in keep_ids]
                edges = [
                    e
                    for e in rows("edges")
                    if e["from_node"] in keep_ids and e["to_node"] in keep_ids
                ]
                chunks = [c for c in rows("chunks") if c["source_node"] in keep_ids]
                provenance = [
                    p for p in rows("ingestion_provenance") if p["node_id"] in keep_ids
                ]
                data = {
                    "nodes": nodes,
                    "edges": edges,
                    "chunks": chunks,
                    "knowledge_sources": rows("knowledge_sources"),
                    "provenance": provenance,
                }
            else:
                data = {
                    "nodes": rows("nodes"),
                    "edges": rows("edges"),
                    "chunks": rows("chunks"),
                    "knowledge_sources": rows("knowledge_sources"),
                    "provenance": rows("ingestion_provenance"),
                }
        data["counts"] = {k: len(v) for k, v in data.items()}
        return data

    def import_graph_data(
        self, data: Dict[str, Any], *, mode: str = "merge", dry_run: bool = False
    ) -> Dict[str, Any]:
        """Import a logical export back into the store.

        ``mode='merge'`` upserts on top of existing data (id collisions update);
        ``mode='replace'`` clears the graph first. ``dry_run=True`` reports the
        plan without writing. Refuses artifacts from a NEWER graph schema than
        this build.
        """
        nodes = data.get("nodes") or []
        edges = data.get("edges") or []
        chunks = data.get("chunks") or []
        sources = data.get("knowledge_sources") or []
        provenance = data.get("provenance") or []

        header = data.get("header") or {}
        incoming_schema = header.get("graph_schema_version")
        if isinstance(incoming_schema, int) and incoming_schema > GRAPH_SCHEMA_VERSION:
            raise ValueError(
                f"Artifact graph_schema_version {incoming_schema} is newer than this "
                f"build ({GRAPH_SCHEMA_VERSION}); refusing to import."
            )

        plan = {
            "mode": mode,
            "nodes": len(nodes),
            "edges": len(edges),
            "chunks": len(chunks),
            "knowledge_sources": len(sources),
            "provenance": len(provenance),
        }
        if dry_run:
            plan["dry_run"] = True
            return plan

        with self._connect() as conn:
            if mode == "replace":
                # Keep replacement imports transactional. The old clear_all()
                # path committed before the import started, so a malformed
                # artifact could leave a cleared or partially rebuilt graph.
                # These deletes roll back with the rest of the import.
                for table in (
                    "local_file_index",
                    "knowledge_sources",
                    "chunks",
                    "edges",
                    "nodes",
                    "vector_embeddings",
                ):
                    conn.execute(f"DELETE FROM {table}")
                if KGStoreV2 is not None:
                    conn.execute("DELETE FROM edges_v2")
                    conn.execute("DELETE FROM nodes_v2")
            for n in nodes:
                self._upsert_node(
                    conn,
                    n["id"],
                    n["type"],
                    n.get("title") or "",
                    summary=n.get("summary") or "",
                    metadata=_safe_loads(n.get("metadata_json")),
                    raw=_safe_loads(n.get("raw_json")),
                )
            for c in chunks:
                self._upsert_chunk(
                    conn,
                    chunk_id=c["id"],
                    source_node=c["source_node"],
                    text=c.get("text") or "",
                    metadata=_safe_loads(c.get("metadata_json")),
                )
            for e in edges:
                e_meta = _safe_loads(e.get("metadata_json")) or {}
                leg_label = e_meta.get("legacy_label")
                if not leg_label:
                    orig = e.get("type") or ""
                    if orig:
                        # preserve whatever label came from export (legacy or canon)
                        leg_label = orig
                self._upsert_edge(
                    conn,
                    e["from_node"],
                    e["to_node"],
                    e["type"],
                    weight=float(e.get("weight") or 1.0),
                    metadata=e_meta,
                    legacy_label=leg_label,
                )
            for s in sources:
                conn.execute(
                    """
                        INSERT OR REPLACE INTO knowledge_sources(
                          id, root_path, os_type, drive_id, label, status, include_ocr,
                          watch_enabled, consent_json, created_at, updated_at, last_scanned_at)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                        """,
                    (
                        s["id"],
                        s["root_path"],
                        s["os_type"],
                        s.get("drive_id"),
                        s.get("label"),
                        s.get("status") or "active",
                        int(s.get("include_ocr") or 0),
                        int(s.get("watch_enabled") or 0),
                        s.get("consent_json") or "{}",
                        s.get("created_at") or _now(),
                        s.get("updated_at") or _now(),
                        s.get("last_scanned_at"),
                    ),
                )
            for p in provenance:
                conn.execute(
                    """
                        INSERT OR REPLACE INTO ingestion_provenance(
                          id, node_id, source_type, source_uri, content_hash, title, pipeline,
                          owner, workspace_id, captured_at, modified_at, embedded, linked,
                          duplicate, agent_used, chunk_count, permissions_json, metadata_json, created_at)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                        """,
                    (
                        p["id"],
                        p["node_id"],
                        p["source_type"],
                        p.get("source_uri"),
                        p.get("content_hash"),
                        p.get("title"),
                        p.get("pipeline") or "import",
                        p.get("owner"),
                        p.get("workspace_id"),
                        p.get("captured_at"),
                        p.get("modified_at"),
                        int(p.get("embedded") or 0),
                        int(p.get("linked") or 0),
                        int(p.get("duplicate") or 0),
                        p.get("agent_used"),
                        int(p.get("chunk_count") or 0),
                        p.get("permissions_json") or "{}",
                        p.get("metadata_json") or "{}",
                        p.get("created_at") or _now(),
                    ),
                )
        plan["imported"] = True
        return plan

    def backup_database(self, dest_path) -> Path:
        """Write a clean, standalone snapshot of the live DB to ``dest_path``.

        Uses ``VACUUM INTO`` (after a full WAL checkpoint) so the snapshot is a
        defragmented, rollback-journal-mode database with no companion -wal/-shm
        — which restores cleanly by a plain file copy. Captures all data incl.
        the vector_embeddings BLOBs.
        """
        dest = Path(dest_path)
        dest.parent.mkdir(parents=True, exist_ok=True)
        if dest.exists():
            dest.unlink()  # VACUUM INTO requires the target to not exist
        # Raw connection, not ``_connect()``: VACUUM cannot run inside a
        # transaction, and ``_connect()`` wraps its block in one.
        conn = self.storage_engine.connect()
        try:
            conn.execute("PRAGMA wal_checkpoint(FULL)")
            conn.execute("VACUUM INTO ?", (str(dest),))
        finally:
            conn.close()
        return dest

File: rust/lattice-core/src/graph_write/provenance.rs
//! Provenance and portability — port of `lattice_brain/graph/provenance.py`.
//!
//! `record_ingestion` is `record_provenance`: an upsert keyed on
//! **`(node, content, source_type, source_uri, pipeline)`** and deliberately
//! *not* on the wall clock. Through 11.0.x the basis included a
//! second-resolution timestamp, which made a record's identity depend on when
//! it happened: re-ingesting unchanged content twice inside one second
//! collapsed onto a single row, and one second later appended a duplicate. That
//! is not an audit trail, it is a race — and it grew this table without bound
//! on every re-scan of an unchanged folder.

use serde_json::{json, Map, Value};

use crate::db::CoreError;

use super::pyaux::{json_of, py_str, sha256_text, truthy};
use super::types::{EdgeSpec, ImportRequest, IngestionRecord, NodeSpec, RebuildRequest};
use super::vectors::RebuildOutcome;
use super::GraphWriter;

/// What `record_provenance` returns.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProvenanceReceipt {
    pub id: String,
    pub node_id: String,
    pub created_at: String,
}

impl ProvenanceReceipt {
    /// `{"id": …, "node_id": …, "created_at": …}`.
    pub fn to_json(&self) -> Value {
        json!({"id": self.id, "node_id": self.node_id, "created_at": self.created_at})
    }
}

/// What `import_graph_data` returns (the plan, plus what happened to it).
#[derive(Debug, Clone)]
pub struct ImportOutcome {
    pub mode: String,
    pub nodes: usize,
    pub edges: usize,
    pub chunks: usize,
    pub knowledge_sources: usize,
    pub provenance: usize,
    pub dry_run: bool,
    pub imported: bool,
    /// The re-derived vector index. `None` on a dry run.
    pub index: Option<RebuildOutcome>,
}

impl ImportOutcome {
    /// The plan body, without the `index` block.
    ///
    /// `index` is composed in Python from the read-side `vector_freshness()`
    /// report, which belongs to the retrieval layer rather than to the write
    /// engine; the rebuild it wraps *is* observable here, because it writes a
    /// `vector_index_operations` row.
    pub fn to_json(&self) -> Value {
        let mut map = Map::new();
        map.insert("mode".into(), json!(self.mode));
        map.insert("nodes".into(), json!(self.nodes));
        map.insert("edges".into(), json!(self.edges));
        map.insert("chunks".into(), json!(self.chunks));
        map.insert("knowledge_sources".into(), json!(self.knowledge_sources));
        map.insert("provenance".into(), json!(self.provenance));
        if self.dry_run {
            map.insert("dry_run".into(), json!(true));
        }
        if self.imported {
            map.insert("imported".into(), json!(true));
        }
        Value::Object(map)
    }
}

impl GraphWriter {
    /// `record_provenance` — where an ingested node came from.
    pub fn record_ingestion(
        &self,
        record: &IngestionRecord,
    ) -> Result<ProvenanceReceipt, CoreError> {
        let now = self.clock.now_iso();
        let basis = format!(
            "{}|{}|{}|{}|{}",
            record.node_id,
            record.content_hash.clone().unwrap_or_default(),
            record.source_type,
            record.source_uri.clone().unwrap_or_default(),
            record.pipeline,
        );
        let prov_id = format!("prov:{}", &sha256_text(&basis)[..24]);
        self.store.with_write_txn(|txn| {
            txn.execute(
                "INSERT OR REPLACE INTO ingestion_provenance( \
                   id, node_id, source_type, source_uri, content_hash, title, pipeline, \
                   owner, workspace_id, captured_at, modified_at, embedded, linked, \
                   duplicate, agent_used, chunk_count, permissions_json, metadata_json, created_at) \
                 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                rusqlite::params![
                    prov_id,
                    record.node_id,
                    record.source_type,
                    record.source_uri,
                    record.content_hash,
                    record.title,
                    record.pipeline,
                    record.owner,
                    record.workspace_id,
                    record.captured_at,
                    record.modified_at,
                    i64::from(record.embedded),
                    i64::from(record.linked),
                    i64::from(record.duplicate),
                    record.agent_used,
                    record.chunk_count,
                    json_of(&record.permissions),
                    json_of(&record.metadata),
                    now,
                ],
            )?;
            Ok(())
        })?;
        Ok(ProvenanceReceipt {
            id: prov_id,
            node_id: record.node_id.clone(),
            created_at: now,
        })
    }

    /// `import_graph_data` — a logical export back into the store.
    ///
    /// `mode="replace"` clears the graph **inside the same transaction** as the
    /// re-import, so a malformed artifact rolls back to the previous graph
    /// rather than leaving a cleared one. The pre-v4 path committed the clear
    /// first, which is exactly how an import could destroy a Brain.
    pub fn import_graph_data(&self, request: &ImportRequest) -> Result<ImportOutcome, CoreError> {
        let nodes = array_of(&request.data, "nodes");
        let edges = array_of(&request.data, "edges");
        let chunks = array_of(&request.data, "chunks");
        let sources = array_of(&request.data, "knowledge_sources");
        let provenance = array_of(&request.data, "provenance");

        if let Some(header) = request.data.get("header").and_then(Value::as_object) {
            if let Some(incoming) = header.get("graph_schema_version").and_then(Value::as_i64) {
                if incoming > super::schema::GRAPH_SCHEMA_VERSION {
                    return Err(CoreError::InvalidRequest(format!(
                        "Artifact graph_schema_version {incoming} is newer than this \
                         build ({}); refusing to import.",
                        super::schema::GRAPH_SCHEMA_VERSION
                    )));
                }
            }
        }

        let mut outcome = ImportOutcome {
            mode: request.mode.clone(),
            nodes: nodes.len(),
            edges: edges.len(),
            chunks: chunks.len(),
            knowledge_sources: sources.len(),
            provenance: provenance.len(),
            dry_run: request.dry_run,
            imported: false,
            index: None,
        };
        if request.dry_run {
            return Ok(outcome);
        }

        let now = self.clock.now_iso();
        self.store.with_write_txn(|txn| {
            if request.mode == "replace" {
                for table in [
                    "local_file_index",
                    "knowledge_sources",
                    "chunks",
                    "edges",
                    "nodes",
                    "vector_embeddings",
                ] {
                    txn.execute_batch(&format!("DELETE FROM {table}"))?;
                }
                txn.execute_batch("DELETE FROM edges_v2; DELETE FROM nodes_v2;")?;
            }
            for node in &nodes {
                let Some(node) = node.as_object() else { continue };
                self.upsert_node(
                    txn,
                    &NodeSpec {
                        id: text_of(node, "id"),
                        node_type: text_of(node, "type"),
                        title: optional_text(node, "title").unwrap_or_default(),
                        summary: optional_text(node, "summary").unwrap_or_default(),
                        metadata: crate::pytext::safe_loads(
                            node.get("metadata_json").and_then(Value::as_str),
                        ),
                        raw: crate::pytext::safe_loads(
                            node.get("raw_json").and_then(Value::as_str),
                        ),
                        owner: None,
                        workspace_id: None,
                        visibility: None,
                    },
                )?;
            }
            for chunk in &chunks {
                let Some(chunk) = chunk.as_object() else { continue };
                self.upsert_chunk(
                    txn,
                    &text_of(chunk, "id"),
                    &text_of(chunk, "source_node"),
                    &optional_text(chunk, "text").unwrap_or_default(),
                    &crate::pytext::safe_loads(chunk.get("metadata_json").and_then(Value::as_str)),
                )?;
            }
            for edge in &edges {
                let Some(edge) = edge.as_object() else { continue };
                let meta =
                    crate::pytext::safe_loads(edge.get("metadata_json").and_then(Value::as_str));
                // Whatever label the export carried is preserved, so two legacy
                // strings that normalize to one canonical type stay two rows.
                let legacy_label = meta
                    .get("legacy_label")
                    .filter(|value| truthy(value))
                    .map(py_str)
                    .or_else(|| optional_text(edge, "type").filter(|value| !value.is_empty()));
                self.upsert_edge(
                    txn,
                    &EdgeSpec {
                        from_node: text_of(edge, "from_node"),
                        to_node: text_of(edge, "to_node"),
                        edge_type: text_of(edge, "type"),
                        weight: edge
                            .get("weight")
                            .and_then(Value::as_f64)
                            .filter(|value| *value != 0.0)
                            .unwrap_or(1.0),
                        metadata: meta,
                        legacy_label,
                    },
                )?;
            }
            for source in &sources {
                let Some(source) = source.as_object() else { continue };
                txn.execute(
                    "INSERT OR REPLACE INTO knowledge_sources( \
                       id, root_path, os_type, drive_id, label, status, include_ocr, \
                       watch_enabled, consent_json, created_at, updated_at, last_scanned_at) \
                     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                    rusqlite::params![
                        text_of(source, "id"),
                        text_of(source, "root_path"),
                        text_of(source, "os_type"),
                        optional_text(source, "drive_id"),
                        optional_text(source, "label"),
                        optional_text(source, "status")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| "active".into()),
                        int_of(source, "include_ocr"),
                        int_of(source, "watch_enabled"),
                        optional_text(source, "consent_json")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| "{}".into()),
                        optional_text(source, "created_at")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| now.clone()),
                        optional_text(source, "updated_at")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| now.clone()),
                        optional_text(source, "last_scanned_at"),
                    ],
                )?;
            }
            for record in &provenance {
                let Some(record) = record.as_object() else { continue };
                txn.execute(
                    "INSERT OR REPLACE INTO ingestion_provenance( \
                       id, node_id, source_type, source_uri, content_hash, title, pipeline, \
                       owner, workspace_id, captured_at, modified_at, embedded, linked, \
                       duplicate, agent_used, chunk_count, permissions_json, metadata_json, created_at) \
                     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
                    rusqlite::params![
                        text_of(record, "id"),
                        text_of(record, "node_id"),
                        text_of(record, "source_type"),
                        optional_text(record, "source_uri"),
                        optional_text(record, "content_hash"),
                        optional_text(record, "title"),
                        optional_text(record, "pipeline")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| "import".into()),
                        optional_text(record, "owner"),
                        optional_text(record, "workspace_id"),
                        optional_text(record, "captured_at"),
                        optional_text(record, "modified_at"),
                        int_of(record, "embedded"),
                        int_of(record, "linked"),
                        int_of(record, "duplicate"),
                        optional_text(record, "agent_used"),
                        int_of(record, "chunk_count"),
                        optional_text(record, "permissions_json")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| "{}".into()),
                        optional_text(record, "metadata_json")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| "{}".into()),
                        optional_text(record, "created_at")
                            .filter(|value| !value.is_empty())
                            .unwrap_or_else(|| now.clone()),
                    ],
                )?;
            }
            Ok(())
        })?;
        outcome.imported = true;
        // `_reindex_after_import`: the write door already embeds inline, so this
        // is a verification in the common case — but "the index is consistent
        // after an import" has to be a guarantee the import makes, not an
        // accident of where the embedding call happens to live. It is also the
        // only path that records the embedder fingerprint.
        outcome.index = Some(self.rebuild_vector_index(&RebuildRequest {
            full: false,
            include_nodes: true,
            include_chunks: true,
        })?);
        Ok(outcome)
    }
}

fn array_of(data: &Map<String, Value>, key: &str) -> Vec<Value> {
    data.get(key)
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default()
}

fn text_of(map: &Map<String, Value>, key: &str) -> String {
    map.get(key).map(py_str).unwrap_or_default()
}

fn optional_text(map: &Map<String, Value>, key: &str) -> Option<String> {
    map.get(key).filter(|value| !value.is_null()).map(py_str)
}

/// `int(row.get(key) or 0)` — a JSON `null`, `false` or absent key is 0.
fn int_of(map: &Map<String, Value>, key: &str) -> i64 {
    match map.get(key) {
        Some(Value::Bool(flag)) => i64::from(*flag),
        Some(Value::Number(number)) => number.as_i64().unwrap_or(0),
        Some(Value::String(text)) => text.parse::<i64>().unwrap_or(0),
        _ => 0,
    }
}
